### ISSUE TITLE
[inductor] [silent incorrectness] `torch.nn.PairwiseDistance(p=2)` outputs incorrect results with eager

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#151198]
### 🐛 Describe the bug

**symptom**: `torch.nn.PairwiseDistance(p=2)` outputs incorrect results
**device backend**: both triton and CPP
**note**: I have used `fp64` as baseline to compare the results
**repro**
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config
import os

config.fallback_random = True
torch.set_grad_enabled(False)
torch.manual_seed(0)


class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.pad = torch.nn.ReflectionPad3d(1)
        self.dist = torch.nn.PairwiseDistance(p=2)

    def forward(self, x):
        x = self.pad(x)
        x = x.view(x.size(0), -1)
        x = torch.chunk(x, 2, dim=1)
        x = self.dist(x[0], x[1])
        return x


model = Model().eval().cuda()


x = torch.randn(2, 3, 4, 4, 4).cuda()

inputs = [x]


def run_test(model, inputs, backend):
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    torch.manual_seed(0)
    output = model(*inputs)
    return output


output = run_test(model, inputs, 'eager')
c_output = run_test(model, inputs, 'inductor')
fp64 = run_test(model.to(dtype=torch.float64), [x.to(dtype=torch.float64)], 'eager')


print(output)
print(c_output)
print(fp64)
print(torch.allclose(output, c_output, 1e-3, 1e-3, equal_nan=True))
print(torch._dynamo.utils.same(output, c_output, fp64))
print(torch.max(torch.abs(output - c_output)))

```

### Error logs

```
tensor([22.9208, 22.6405], device='cuda:0')
tensor([23.1078, 21.4387], device='cuda:0')
tensor([22.9208, 22.6405], device='cuda:0', dtype=torch.float64)
False
E0414 11:20:47.280000 958741 site-packages/torch/_dynamo/utils.py:2930] RMSE (res-fp64): 0.86004, (ref-fp64): 0.00000 and shape=torch.Size([2]). res.dtype: torch.float32, multiplier: 3.000000, tol: 0.000100, use_larger_multiplier_for_smaller_tensor: 0
False
tensor(1.2018, device='cuda:0')

```

### Versions

nightly 20250414

**Comment by bdhirsh** [comment]
I can repro this with `backend="inductor"` but not `backend="aot_eager_decomp_partition"`

**Comment by eellison** [comment]
@yf225 , any updates ?

**Comment by yf225** [comment]
Some more findings:
- The numerical issue doesn't repro with Inductor post-grad graph -> the bug is either in Inductor lowering from FX IR to scheduler IR, or in scheduler IR passes in scheduler.py.
- I further tried disabling all scheduler IR passes, the numerical issue still exists -> the bug is likely in the Inductor lowering from FX IR to scheduler IR.

**Comment by yf225** [comment]
This is a smaller repro with `ReflectionPad1d` or `ReplicationPad1d`:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config
import os

config.fallback_random = True
torch.set_grad_enabled(False)
torch.manual_seed(0)


class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.pad = torch.nn.ReflectionPad1d(1)  # this fails
        # self.pad = torch.nn.ReplicationPad1d(1)  # this also fails
        # self.pad = torch.nn.CircularPad1d(1)  # this works
        # self.pad = torch.nn.ZeroPad1d(1)  # this works

    def forward(self, x):
        x = self.pad(x)
        x = x.view(x.size(0), -1)
        x = torch.chunk(x, 2, dim=1)
        return torch.sum(x[0] - x[1])


model = Model().eval().cuda()


x = torch.randn(2, 3, 4).cuda()

inputs = [x]


def run_test(model, inputs, backend):
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    torch.manual_seed(0)
    output = model(*inputs)
    return output


output = run_test(model, inputs, 'eager')
c_output = run_test(model, inputs, 'inductor')
fp64 = run_test(model.to(dtype=torch.float64), [x.to(dtype=torch.float64)], 'eager')


print(output)
print(c_output)
print(fp64)
print(torch.allclose(output, c_output, 1e-3, 1e-3, equal_nan=True))
print(torch._dynamo.utils.same(output, c_output, fp64))
print(torch.max(torch.abs(output - c_output)))
```

**Comment by yf225** [comment]
Briefly discussed with @isuruf, it seems the lowering is rewriting rewrite `abs(abs(k-1)-3)` to`abs(k-4)` which is wrong because it ignores the fact that k is constrained to 0..5 .

Also by setting inductor config `constant_and_index_propagation=False`, the numerical issue goes away, suggesting the issue is in IndexPropagation. I will continue to look into that path.

**Comment by isuruf** [comment]
Fixed in https://github.com/pytorch/pytorch/pull/152993

**Comment by yf225** [comment]
Can confirm that @isuruf 's PR fixes this issue

**Comment by eellison** [comment]
Thank you @isuruf and @yf225 ! Also, this is a great example of a very minimal repro that exposes an edge case bug, so thank you @shaoyuyoung !
